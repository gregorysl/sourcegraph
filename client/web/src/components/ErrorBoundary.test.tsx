import * as sentry from '@sentry/browser'
import React from 'react'
import renderer from 'react-test-renderer'
import sinon from 'sinon'

import { HTTPStatusError } from '@sourcegraph/shared/src/backend/fetch'

import { ErrorBoundary } from './ErrorBoundary'

jest.mock('mdi-react/ErrorIcon', () => 'ErrorIcon')
jest.mock('mdi-react/ReloadIcon', () => 'ReloadIcon')

const ThrowError: React.FunctionComponent = () => {
    throw new Error('x')
}

/** Throws an error that resembles the Webpack error when chunk loading fails.  */
const ThrowChunkError: React.FunctionComponent = () => {
    throw new Error('Loading chunk 123 failed.')
}

const ThrowHTTPStatusError: React.FunctionComponent<{ status?: number }> = ({ status = 500 }) => {
    const errorResponse = new Response(null, { status })
    throw new HTTPStatusError(errorResponse)
}

describe('ErrorBoundary', () => {
    test('passes through if non-error', () =>
        expect(
            renderer
                .create(
                    <ErrorBoundary location={null}>
                        <ThrowError />
                    </ErrorBoundary>
                )
                .toJSON()
        ).toMatchSnapshot())

    test('renders error page if error', () =>
        expect(
            renderer
                .create(
                    <ErrorBoundary location={null}>
                        <span>hello</span>
                    </ErrorBoundary>
                )
                .toJSON()
        ).toMatchSnapshot())

    test('renders reload page if chunk error', () =>
        expect(
            renderer
                .create(
                    <ErrorBoundary location={null}>
                        <ThrowChunkError />
                    </ErrorBoundary>
                )
                .toJSON()
        ).toMatchSnapshot())

    test('Sentry should capture HttpStatusError expect Server response error (5xx)', () => {
        const sentryCaptureEx = sinon.stub(sentry, 'captureException')

        expect(
            renderer
                .create(
                    <ErrorBoundary location={null}>
                        <ThrowHTTPStatusError status={500} />
                    </ErrorBoundary>
                )
                .toJSON()
        ).toMatchSnapshot()

        sinon.assert.notCalled(sentryCaptureEx)
        sentryCaptureEx.reset()

        expect(
            renderer
                .create(
                    <ErrorBoundary location={null}>
                        <ThrowHTTPStatusError status={400} />
                    </ErrorBoundary>
                )
                .toJSON()
        ).toMatchSnapshot()

        sinon.assert.calledOnce(sentryCaptureEx)
        expect(sentryCaptureEx.getCall(0).args[0]).toBeInstanceOf(HTTPStatusError)
        expect(sentryCaptureEx.getCall(0).args[0]).toHaveProperty('status', 400)

        sentryCaptureEx.restore()
    })
})
