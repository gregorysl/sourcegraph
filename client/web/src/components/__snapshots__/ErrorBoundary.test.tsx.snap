// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Sentry should capture HttpStatusError expect Server response error (5xx) 1`] = `
<div
  className="heroPage defaultPadding"
>
  <div
    className="icon"
  >
    <ErrorIcon />
  </div>
  <div
    className="title"
  >
    Error
  </div>
  <div
    className="subtitle"
    data-testid="hero-page-subtitle"
  >
    <div
      className="container"
    >
      <p>
        Sourcegraph encountered an unexpected error. If reloading the page doesn't fix it, contact your site admin or Sourcegraph support.
      </p>
      <p>
        <code
          className="text-wrap"
        >
          Request to  failed with 500 Internal Server Error
        </code>
      </p>
    </div>
  </div>
</div>
`;

exports[`ErrorBoundary Sentry should capture HttpStatusError expect Server response error (5xx) 2`] = `
<div
  className="heroPage defaultPadding"
>
  <div
    className="icon"
  >
    <ErrorIcon />
  </div>
  <div
    className="title"
  >
    Error
  </div>
  <div
    className="subtitle"
    data-testid="hero-page-subtitle"
  >
    <div
      className="container"
    >
      <p>
        Sourcegraph encountered an unexpected error. If reloading the page doesn't fix it, contact your site admin or Sourcegraph support.
      </p>
      <p>
        <code
          className="text-wrap"
        >
          Request to  failed with 400 Bad Request
        </code>
      </p>
    </div>
  </div>
</div>
`;

exports[`ErrorBoundary passes through if non-error 1`] = `
<div
  className="heroPage defaultPadding"
>
  <div
    className="icon"
  >
    <ErrorIcon />
  </div>
  <div
    className="title"
  >
    Error
  </div>
  <div
    className="subtitle"
    data-testid="hero-page-subtitle"
  >
    <div
      className="container"
    >
      <p>
        Sourcegraph encountered an unexpected error. If reloading the page doesn't fix it, contact your site admin or Sourcegraph support.
      </p>
      <p>
        <code
          className="text-wrap"
        >
          x
        </code>
      </p>
    </div>
  </div>
</div>
`;

exports[`ErrorBoundary renders error page if error 1`] = `
<span>
  hello
</span>
`;

exports[`ErrorBoundary renders reload page if chunk error 1`] = `
<div
  className="heroPage defaultPadding"
>
  <div
    className="icon"
  >
    <ErrorIcon />
  </div>
  <div
    className="title"
  >
    Error
  </div>
  <div
    className="subtitle"
    data-testid="hero-page-subtitle"
  >
    <div
      className="container"
    >
      <p>
        Sourcegraph encountered an unexpected error. If reloading the page doesn't fix it, contact your site admin or Sourcegraph support.
      </p>
      <p>
        <code
          className="text-wrap"
        >
          Loading chunk 123 failed.
        </code>
      </p>
    </div>
  </div>
</div>
`;
